tensorly/tenalg/proximal.py

```diff
@@ -91,4 +91,316 @@ def procrustes(matrix):
+
+    It features two accelerations: an early stop stopping criterion, and a
+    complexity averaging between precomputations and loops, so as to use large
+    precomputations several times.
+
+    This function is made for being used repetively inside an
+    outer-loop alternating algorithm, for instance for computing nonnegative
+    matrix Factorization or tensor factorization.
+
+    Parameters
+    ----------
+    UtM: r-by-n array
+        Pre-computed product of the transposed of U and M, used in the update rule
+    UtU: r-by-r array
+        Pre-computed product of the transposed of U and U, used in the update rule
+    in_V: r-by-n initialization matrix (mutable)
+        Initialized V array
+        By default, is initialized with one non-zero entry per column
+        corresponding to the closest column of U of the corresponding column of M.
+    maxiter: Postivie integer
+        Upper bound on the number of iterations
+        Default: 500
+    delta : float in [0,1]
+        early stop criterion, while err_k > delta*err_0. Set small for
+        almost exact nnls solution, or larger (e.g. 1e-2) for inner loops
+        of a PARAFAC computation.
+        Default: 10e-8
+    sparsity_coefficient: float or None
+        The coefficient controling the sparisty level in the objective function.
+        If set to None, the problem is solved unconstrained.
+        Default: None
+    nonzero_row: boolean
+        True if the lines of the V matrix can't be zero,
+        False if they can be zero
+        Default: False
+
+    Returns
+    -------
+    V: array
+        a r-by-n nonnegative matrix \approx argmin_{V >= 0} ||M-UV||_F^2
+    eps: float
+        number of loops authorized by the error stop criterion
+    cnt: integer
+        final number of update iteration performed
+    rho: float
+        number of loops authorized by the time stop criterion
+
+    References
+    ----------
+    [1]: N. Gillis and F. Glineur, Accelerated Multiplicative Updates and
+    Hierarchical ALS Algorithms for Nonnegative Matrix Factorization,
+    Neural Computation 24 (4): 1085-1105, 2012.
+
+    [2] J. Eggert, and E. Korner. "Sparse coding and NMF."
+    2004 IEEE International Joint Conference on Neural Networks
+    (IEEE Cat. No. 04CH37541). Vol. 4. IEEE, 2004.
+
+    """
+
+    r, n = tl.shape(UtM)
+    if in_V is None:  # checks if V is empty
+        V = tl.solve(UtU, UtM)
+
+        V[V < 0] = 0
+        # Scaling
+        scale = tl.sum(UtM * V) / tl.sum(
+            UtU * tl.dot(V, tl.transpose(V)))
+        V = tl.dot(scale, V)
+    else:
+        V = in_V
+
+    rho = 1
+    eps0 = 0
+    cnt = 1
+    eps = 1
+
+
+    while eps >= delta * eps0 and cnt <= 1 + 0.5* rho and cnt <= maxiter:
+        nodelta = 0
+        for k in range(r):
+
+            if UtU[k, k] != 0:
+                if sparsity_coefficient != None: # Modifying the objective function for sparsification
+                    if tl.get_backend() == 'pytorch':
```

Comment:
It's best to avoid backend specific code.

Reply:
I removed all the backend options